### PR TITLE
fix(spawner): rewrite to use createVat, strip down to bare minimum API

### LINKED ISSUE #1343: spawner should create dynamic vats, rather than evaluating code in the spawner's own vat
We should move the `spawner` package to put code into new dynamic vats, just like the new split-Zoe does.

This will unblock #1341.

### PR DESCRIPTION
**Pull request by warner**
closes #1343

### CODE REVIEW

**Comment by warner** [comment]
Could y'all double-check that:

* the README accurately describes what we're still using the spawner for
* I think we're only accepting a single argument (so `install(x)`, not `install(x,y)`) to be passed through to the `startFn`.. is that ok? Will that be surprising?
* Does the way that a runaway vat is signalled good enough? The previous test checked that the error was more detailed ("allocate meter exceeded") that the simple "vat terminated" you get now.
* Did I rip out too much? There seemed to be a lot of zoe-ish contract support code that I think is no longer needed, so I was pretty aggressive with the delete key.

cc @katelynsills would love to have your opinion too

**Comment by warner** [comment]
Oh, I think I renamed something (`makeContractHost`) that I should have left alone.

**Comment by warner** [comment]
Oh, right, so I renamed `makeContractHost` to `makeSpawner` because the API changed: it now requires `vatAdminSvc` as an argument, so it didn't make sense to keep the old name around.

The new implementation retains the `install`/`spawn` API, so I think the deploy script itself doesn't have to change, but any *tests* in dapps that thought they could build the spawner themselves: those need to change.

I notice that most of the Dapp CI tests are passing. Is there something different about the getting-started integration test that makes it more sensitive?

Is there something I can do to spawner to let that pass, or do we need to change the getting-started code in some way?

**Comment by warner** [comment]
And I don't know why `test-cosmic-swingset` failed.

**Comment by michaelfig** [comment]
> And I don't know why `test-cosmic-swingset` failed.

I'll look into it now.

**Comment by michaelfig** [comment]
> > And I don't know why `test-cosmic-swingset` failed.
> 
> I'll look into it now.

Please try running `rg makeContractHost agoric-sdk`.  I'm pretty sure that no dapps or tests outside of `spawner` use `makeContractHost`, but there are some references in `cosmic-swingset` that will need to be updated.

But yes, please do the rename.

**Comment by michaelfig** [comment]
> I notice that most of the Dapp CI tests are passing. Is there something different about the getting-started integration test that makes it more sensitive?

It runs `cosmic-swingset`.

**Comment by warner** [comment]
oops, yes, I meant to fix up the call from cosmic-swingset and forgot.. pulling from review until I get that done

**Comment by warner** [comment]
In addition to not updating cosmic-swingset, I also forgot to give `{ console, assert }` as endowments to the imported bundle, which failed when run against a real bundle (such as one that imports `@agoric/assert` and looks for an `assert` on the globals). I enhanced the spawner test to exercise both that and `console`, and removed even more code in the process.

I was able to test dapp-fungible-faucet against this, which I *think* exercises as much of the spawner as we intend to support.

**Comment by warner** [comment]
Oh, one other note: I had to add a `yarn build` step to `packages/spawner` to create the bundle that gets installed in each new dynamic vat (the code that does `importBundle` of the default-exporting bundle). Any changes to the source file in `packages/spawner/src/vat-spawned.js` will necessitate a new `yarn build`, as will updating your git checkout to the tree after this PR lands.

**Comment by warner** [comment]
@michaelfig I'm not seeing any obvious errors in the getting-started test output. Do you think maybe it's just taking longer than the timeout allows?

File: packages/spawner/scripts/build-bundle-spawn.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      if (err) throw err;
+    });
+    fs.writeFileSync(outputPath, `export default ${JSON.stringify(bundle)};`);
+  });
+}
+
+async function main() {
+  const contractFilename = `${__dirname}/../src/vat-spawned.js`;
+  const outputPath = `${__dirname}/../bundles/bundle-spawn.js`;
+  await writeSourceBundle(contractFilename, outputPath);
```
Comment (Chris-Hibbert) [inline packages/spawner/scripts/build-bundle-spawn.js:19]: (Not an issue for this PR)

That seems like a security hole: `dirname` is a global, and you can navigate upwards from it?  Is this not available from SES?

**Review by Chris-Hibbert — APPROVED** [review]
Thanks for the cleanup!  My one question is about security in SES, and unrelated to this PR, I think.

**Comment by katelynsills** [comment]
Just did a cursory look, but seems great!

**Comment by warner** [comment]
I *think* that integration test is merely something taking longer than it did before (such as creating a new dynamic vat, which takes 10-20s until we get switched over to XS). I'm checking now, but if so, the fix will just be to increase a 60s timeout within the integration test.

**Comment by warner** [comment]
Yep, the `agoric deploy ./contract/deploy.js ./api/deploy.js` step in the integration test had a 60s timeout. On my local machine, this used to take 47s, but with the extra dynamic vats created by the new spawner, this increased to 76s and exceeded the timeout. I raised it to 120s and it seems to pass reliably.

This will become a non-issue when we move to XS, and dynamic vat creation should speed up by maybe 8-10x.

File: packages/spawner/scripts/build-bundle-spawn.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import 'ses';
+import fs from 'fs';
+import process from 'process';
+import bundleSource from '@agoric/bundle-source';
+
+async function writeSourceBundle(contractFilename, outputPath) {
+  await bundleSource(contractFilename).then(bundle => {
+    // TODO: fix
+    // @ts-ignore mkdirSync believes it only accepts 2 arguments.
+    fs.mkdirSync(`${__dirname}/../bundles`, { recursive: true }, err => {
```
Comment (michaelfig) [inline packages/spawner/scripts/build-bundle-spawn.js]: `fs.mkdirSync` *does* only accept 2 arguments, and it throws on error.  The callback is only used for `fs.mkdir`.

**Review by michaelfig — APPROVED** [review]
LGTM!

Thanks for this.

File: packages/spawner/scripts/build-bundle-spawn.js (hunk 2 above)
Comment (warner) [reply packages/spawner/scripts/build-bundle-spawn.js]: ah, I copied that from `zoe/scripts/build-zcfBundle.js`.. I'll make a PR to fix it over there too.

**Pull request merged**